src/flasher_spi.c

```diff
@@ -537,6 +537,235 @@ NETX_CONSOLEAPP_RESULT_T spi_flash(const FLASHER_SPI_FLASH_T *ptFlashDescription
+		return retVal;
+	}
+
+	/* Get sector size, set a cap if it exceeds the read buffer size */
+	unsigned long sectorSizeBytes = attributes.ulSectorPages * attributes.ulPageSize;
+	if(sectorSizeBytes > SPI_BUFFER_SIZE)
+	{
+		sectorSizeBytes = SPI_BUFFER_SIZE;
+	}else if(sectorSizeBytes == 0){
+		uprintf("! Could not get sector size. Page size: %d, Sector pages: %d\n", attributes.ulPageSize, attributes.ulSectorPages);
+		return NETX_CONSOLEAPP_RESULT_ERROR;
+	}
+
+	/* This assumes the erase entries are sorted (which they should be) */
+	unsigned long maxEraseSize = eraseTypes[nrEraseOps-1].Size;
+
+	/* Maximum number of chunks we consider at once, limited by the largest erase instruction */
+	unsigned long nrSectorsInLargestErase = maxEraseSize/sectorSizeBytes;
+
+	/* Number of bytes we need in the sector bitmap (+1 to account for rounding) */
+	unsigned int bitPerMapEntry = (sizeof(unsigned int)*8);
+	unsigned int bitmapSize = nrSectorsInLargestErase / bitPerMapEntry + 1;
+	
+	/* Bitmap to store the erase status by sector */
+	#define SPI_ERASE_BITMAP_ENTRY_CNT 32 /* This should be more than enough, allows erase commands of up to 1024 sectors */
+	unsigned int sectorBitmap[SPI_ERASE_BITMAP_ENTRY_CNT];
+	for (unsigned int i = 0; i < SPI_ERASE_BITMAP_ENTRY_CNT; i++)
+	{
+		sectorBitmap[i] = 0;
+	}
+	
+	if(bitmapSize > SPI_ERASE_BITMAP_ENTRY_CNT){
+		uprintf("! Remembering too many memory sections required for largest erase command, falling back to normal erase\n");
+		retVal = spi_erase(ptFlashDescription, ulStartAdr, ulEndAdr);
+		return retVal;
+	}
+
+	/* For the first iteration, align the sector map with sectors of largest erase command */
+	unsigned long currAdr = ulStartAdr;
+	unsigned long currSector = 0;
+	currSector = (currAdr % maxEraseSize)/sectorSizeBytes;
+
+	/* Go through entire erase area */
+	progress_bar_init(ulEndAdr-ulStartAdr);
+	while(currAdr < ulEndAdr)
+	{
+		/* Read a set of sectors */
+		while(currSector < nrSectorsInLargestErase)
+		{
+			if(currAdr >= ulEndAdr) break; /* Stop reading if we reached the end */
+			retVal = Drv_SpiReadFlash(ptFlashDescription, currAdr, pucSpiBuffer, sectorSizeBytes);
+			if(retVal != NETX_CONSOLEAPP_RESULT_OK)
+			{
+				uprintf("! Could not read the flash at address 0x%x\n", currAdr);
+				return retVal;
+			}
+
+			if(!sectorIsEmpty(sectorSizeBytes, pucSpiBuffer))
+			{
+				bitmapWriteBit(1, currSector, sectorBitmap);
+			}else
+			{
+				bitmapWriteBit(0, currSector, sectorBitmap);
+			}
+			currAdr = currAdr + sectorSizeBytes;
+			currSector = currSector + 1;
+		}
+
+		/* Find optimal erase pattern for read sectors */
+		/* Check each erase command, starting with the largest */
+		for(int eraseCmd = (int) nrEraseOps-1; eraseCmd >= 0; eraseCmd--)
+		{
+			unsigned int nrSectorsEraseable = eraseTypes[eraseCmd].Size/sectorSizeBytes;
+			/* Check if we found a page erase command, we only erase by sector */
+			if(nrSectorsEraseable < 1)
+			{
+				continue;
+			}
```

Comment:
This means that the routine cannot erase with page granularity even on a flash device that supports page erase.
Why is the logic based on sector erase and not on the smallest erase block size?

The normal SPI erase routine, spi_erase_with_progress, can erase with page granularity.
However, it won't do so normally, because spi_getEraseArea always extends the area to be erased to sector bounaries.
So maybe it's no problem that we can't erase single pages.

Reply:
"My biggest concern is that we're only gaining a marginal improvement" - I wasn't thinking about a speed increase, but rather about being able to erase smaller areas, but as said, since get_erase_area expands the area to the next sector boundaries anyway and we didn't need page-granularity so far, I don't think it's important.